2500: Add support for default face attributes

GoaLitiuM opened the pull request:
This PR adds support for default face attributes for new brushes by defining the default values in game configuration file. The default options and all the fields are optional so only the needed default attributes can be added to the game config when needed.

example GameConfig.cfg: 
```json
{
   "faceattribs": {
      "defaults": {
         "textureName": "defaultTexture",
         "offset": [0, 0],
         "scale": [0.5, 0.5],
         "rotation": 0,
         "surfaceContents": ["solid"],
         "surfaceFlags": ["skip"],
         "surfaceValue": 0,
         "color": "1.0 1.0 1.0 1.0",
      }
   }
}
```

Closes #2500.

GoaLitiuM commented:
The implementation part needs some feedback. I think I'm quite happy with the parsing part but I wasn't completely sure where the logic for reading and applying the attributes should be placed.

GoaLitiuM commented on common/src/IO/GameConfigParser.cpp:
```diff
@@ -228,6 +229,54 @@ namespace TrenchBroom {
+            if (!value["surfaceContents"].null()) {
+                for (size_t i = 0; i < value["surfaceContents"].length(); ++i) {
+                    auto name = value["surfaceContents"][i].stringValue();
+                    defaults.surfaceContents += contentFlags.flagValue(name);
```
Oopsie, this should use  `|=` and not `+=`

eGax commented:
Can I define scale for each texture on an individual basis as well?

"brushface": [
            {
                "name": "chain",
                "attribs": [ "transparent" ],
                "scale": [4., 4.0],
                "match": "texture",
                "pattern": "chainlink"
            },

That would be AWESOME!

eGax commented:
Let me rephrase my last question: Can I define these default attributes to an individual texture as well?

Anything defined in defaults like that IMO would be great to have available in user-defined smart tags (brush/brushface) defined in the GameConfig.cfg for individual textures.

GoaLitiuM commented:
>Can I define these default attributes to an individual texture as well?
Currently, no, but this would probably require some kind of tracking of user changed attributes so we can safely change the attributes without overriding user set values, unless it's actually desired, which makes this even more trickier to implement.

eGax commented:
But isn't your faceattribs additions being defined in the GameConfig.cfg?

{
   "faceattribs": {
      "defaults": {
         "textureName": "defaultTexture",
         "offset": [0, 0],
         "scale": [0.5, 0.5],
         "rotation": 0,
         "surfaceContents": ["solid"],
         "surfaceFlags": ["skip"],
         "surfaceValue": 0,
         "color": "1.0 1.0 1.0 1.0",
      }
   }
}

GoaLitiuM commented:
They are currently defined there (not sure if it's the right place for it), but it can be changed if it doesn't make sense to be there. For me that was the logical place for it.

Maybe this should be implemented by using the tag system instead? This way the current global override behavior could be imitated by adding a brush tag matching against worldspawn type entities, and this could be expanded to match against textures as well with brushface matching.

eGax commented:
It makes sense for them to be in the GameConfig.cfg to me as well.

If that's the case then if it were expanded to match against textures as well with brushface matching then any individual defined smart tags located after your default additions that contain a different value from the defaults would override them.

kduske commented:
Per texture defaults are an entirely different matter and could be solved via tags, as they are updated dynamically. But that should not be part of this PR.

kduske commented:
@GoaLitiuM let me know when this PR is ready to be reviewed

ericwa commented on common/src/Model/GameConfig.cpp:
```diff
@@ -165,11 +166,54 @@ namespace TrenchBroom {
+                                                             const vm::vec2f& i_offset,
+                                                             const vm::vec2f& i_scale,
+                                                             const float& i_rotation,
+                                                             const int& i_surfaceContents,
```
minor nitpick, these should be just `float` instead of `const float&` and `int` instead of `const int&` (other types should stay as const reference)

GoaLitiuM commented:
@kduske This should be ready now so go ahead.

GoaLitiuM commented on common/src/View/CreateSimpleBrushTool.cpp:
```diff
@@ -33,7 +34,8 @@ namespace TrenchBroom {
-            const Model::BrushBuilder builder(document->world(), document->worldBounds());
+            auto game = document->game();
+            const Model::BrushBuilder builder(document->world(), document->worldBounds(), game->defaultFaceAttribs());
             updateBrush(builder.createCuboid(bounds, document->currentTextureName()));
```
Default textures are not being used at all because `BrushBuilder::createCuboid` and `BrushBuilder::createBrush` does not have texture name as an optional parameter.

GoaLitiuM commented on common/src/Model/GameConfig.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "GameConfig.h"
 
 #include "Ensure.h"
+#include "Model/BrushFace.h"
```
Quite sizable header file needed for `BrushFace::NoTextureName`, ugh.

kduske commented:
Ok, I'll look at this tomorrow. Meanwhile, I have dropped a big change onto master that broke this branch, but it should be easy to fix.

GoaLitiuM commented:
Possibly unrelated but after the changes in master branch, TB seems to crash quite often in `Hit::target()` as soon as I hover over an entity or attempt to create a brush.

```
 	KernelBase.dll!76cf94f2()	Unknown
 	KernelBase.dll![Frames below may be incorrect and/or missing, no symbols loaded for KernelBase.dll]	Unknown
 	[External Code]	
>	TrenchBroom-nomanual.exe!TrenchBroom::Model::Hit::target<TrenchBroom::Model::Node *>() Line 67	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::Model::hitToNode(const TrenchBroom::Model::Hit & hit) Line 32	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::Model::HitQuery::visible(const TrenchBroom::Model::Hit & hit) Line 156	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::Model::HitQuery::first() Line 113	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::SelectionTool::firstHit(const TrenchBroom::View::InputState & inputState, unsigned __int64 type) Line 219	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::SelectionTool::doSetRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 413	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolControllerBase<TrenchBroom::View::NoPickingPolicy,TrenchBroom::View::NoKeyPolicy,TrenchBroom::View::MousePolicy,TrenchBroom::View::MouseDragPolicy,TrenchBroom::View::RenderPolicy,TrenchBroom::View::NoDropPolicy>::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 495	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 139	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolChain::setRenderOptions(const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 142	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolBox::setRenderOptions(TrenchBroom::View::ToolChain * chain, const TrenchBroom::View::InputState & inputState, TrenchBroom::Renderer::RenderContext & renderContext) Line 248	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::ToolBoxConnector::setRenderOptions(TrenchBroom::Renderer::RenderContext & renderContext) Line 113	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::MapViewBase::doRender() Line 868	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::RenderView::render() Line 210	C++
 	TrenchBroom-nomanual.exe!TrenchBroom::View::RenderView::paintGL() Line 164	C++
 	[External Code]	
 	TrenchBroom-nomanual.exe!TrenchBroom::View::TrenchBroomApp::notify(QObject * receiver, QEvent * event) Line 509	C++
 	[External Code]	
 	TrenchBroom-nomanual.exe!main(int argc, char * * argv) Line 66	C++
 	TrenchBroom-nomanual.exe!WinMain(HINSTANCE__ * __formal, HINSTANCE__ * __formal, char * __formal, int __formal) Line 104	C++
 	[External Code]	
```

kduske commented:
Please fix the errors reported by the Travis Linux build.

kduske commented on common/src/View/MapView2D.cpp:
```diff
@@ -201,7 +203,7 @@ namespace TrenchBroom {
             assert(!selectionBrushes.empty());
 
-            const Model::BrushBuilder brushBuilder(document->world(), worldBounds);
+            const Model::BrushBuilder brushBuilder(document->world(), worldBounds, game->defaultFaceAttribs());
```
For this brush, you don't need to apply the defaults at all, it will be removed immediately.

GoaLitiuM commented on common/src/IO/GameConfigParser.cpp:
```diff
@@ -228,6 +229,54 @@ namespace TrenchBroom {
+            if (!value["textureName"].null()) {
+                defaults.textureName = value["textureName"].stringValue();
+            }
+            if (value["offset"].length() == 2) {
```
Null value types always return 0 for length, so it's not really needed.

kduske replied:
Ah, I see.

GoaLitiuM commented:
I got rid of the unneeded configuration structure and replaced it with ``BrushFaceAttributes``, fixed the ``FaceAttribsConfig`` `==` operator and bunch of other minor changes.

GoaLitiuM commented on common/src/IO/GameConfigParser.cpp:
```diff
@@ -228,6 +229,54 @@ namespace TrenchBroom {
             return flags;
         }
 
+        Model::FaceAttribsDefaultsConfig GameConfigParser::parseFaceAttribsDefaults(const EL::Value& value, const Model::FlagsConfig& surfaceFlags, const Model::FlagsConfig& contentFlags) const {
```
Agreed, while it looks consistent with the other structures, it just adds extra maintenance.

kduske commented on common/src/Model/GameConfig.h:
```diff
@@ -22,6 +22,7 @@
 
 #include "Color.h"
 #include "IO/Path.h"
+#include "Model/Brushface.h"
```
I think this can be removed now?

kduske commented:
Thanks!
